### PR TITLE
[Feature] JavaScript Addons Support

### LINKED ISSUE #7959: [Feature] JavaScript Addons support for AnkiDroid
**Is your feature request related to a problem? Please describe.**
For Anki, addons can be made using Python. But for AnkiDroid JavaScript may be option to implement some of the addons. There are already some API have been exposed in AnkiDroid JS API. But deck developer have to tell or guide user to add ```HTML/CSS/JS```. So this is idea for having js addons. 

The addons made in javascript so it will be used in Anki Ecosystem as Anki Desktop and Anki Mobile use webview for reviewer. But the AnkiDesktop and AnkiMobile have not exposed the API like AnkiDroid JS API. So, in #7958 PR there are implementation for containing the addons to AnkiDroid only. It can be extented for AnkiDesktop and AnkiMobile if both have API.

But GOAL here is to have addon support so that all deck should share same ```HTML/CSS/JS```  like addon for AnkiDroid. The addons will use AnkiDroid JS API.

**Describe the solution you'd like**
#7958 . I am implementing this Feature but it requires some discussion and possible changes to make this feature more useful.

**Describe alternatives you've considered**
The alternative solution will be copying pasting the code to Card template manually.

**Additional context**
The workflow will be
1. User will download the addon ```.adaddon``` (AnkiDroid Addons) or ```.jsaddon``` (JavaScript Addon) files
2. Install using in app option (to be implemented)
3. Enable the addons, so that the contents of addons will be added to reviewer during review. 

捕获

```
Example of some Anki Addons that can be implemented
1. Progress Bar
2. Custom buttons at bottom
3. Card layout changes
4. Cards info during review
5. Clickable tag
etc...
```
The more API get exposed the more features will addon will have.

### PR DESCRIPTION
**Pull request by infinyte7**
## Purpose / Description
Using AnkiDroid JS API a deck can be extended to implement features like progress bar or custom card layouts. But it requires user to manually add the ```html/css/js``` to card templates. This is implementation for a way to inject code to all deck templates when addons get activated.

## Fixes
Fixes #7959 

## Approach
1. Download addons from https://www.npmjs.com/search?q=keywords:ankidroid-js-addon inside the app
2. In app there are implementation to check for valid addons for AnkiDroid
It will check following string in latest package.json
```
{
    "name": "ankidroid-js-addon-progress-bar",
    "version": "1.0.0",
    "author": "https://github.com/infinyte7",
    "homepage": "https://github.com/infinyte7/Anki-Custom-Card-Layout",
    "ankidroid_js_api": "0.0.1",
    "addon_type": "reviewer",
    "keywords": ["ankidroid-js-addon"]
}
```
There isValidPackageJson boolean function will check for above values if present then the addons is valid.

3. The ```enabled``` addons' ```index.js``` content wrapped in ```<script>``` tag will be added to ```fillFlashCard()``` in ```AbstractFlashcardViewer.java```

There are two types of addon ```reviewer``` and ```note_editor```. In this PR, reviewer addon is implemented.

Idea for ```note_editor``` addon:
The javascript function can be called using https://github.com/evgenyneu/js-evaluator-for-android/#call-a-javascript-function . So, in note editor multiple buttons will get added to toolbar with tag containing addons' name. When the toolbar button pressed, it will take ```selected text``` from FieldEditText and pass to jsEvaluate java function. The jsEvaluate java function will get tag of clicked toolbar button and open js file with name equal to tag of button and pass the selected text to addon's function implementation. The addon's js function will return the manipulated string to FieldEditText in note editor. Even if other implementation can be done without selected text as defined in js file in addon's folder. This will implemented in separated PR.

## How Has This Been Tested?
Tested on emulator 

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration (SDK version(s), emulator or physical, etc)

## Learning (optional, can help others)
https://developer.android.com/guide/topics/ui/layout/recyclerview

## Checklist
_Please, go through these checks before submitting the PR._

- [ x ] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [ x ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ x ] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ x ] You have commented your code, particularly in hard-to-understand areas
- [ x ] You have performed a self-review of your own code
- [ x ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)

### CODE REVIEW

**Comment by infinyte7** [comment]
This is first commit and need help in implementing way to store the activated addons status.

File: AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        Switch addonActivate;
+        ImageButton addonDelete;
+        ImageButton addonInfo;
+        public AddonsViewHolder(@NonNull View itemView) {
+            super(itemView);
+            addonsTextView = (TextView) itemView.findViewById(R.id.addon_name);
+            addonActivate = (Switch) itemView.findViewById(R.id.activate_addon);
+            addonDelete = (ImageButton) itemView.findViewById(R.id.delete_addon);
+            addonInfo = (ImageButton) itemView.findViewById(R.id.addon_info);
+
```
Comment (infinyte7) [inline AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:52]: How can I store the values for the activated addons? So, content.html from activated addons will be added to ```fillFlashcard()``` in ```AbstractFlashcardViewer.java```.

**Comment by Arthur-Milchior** [comment]
I really like the general idea.
Given that there is no monkey patching, I hope that this will ensure that add-ons live together better in ankidroid than in anki (even if it became better in anki the two last years)
I hope and fear it should be cool to have a place to list all add-ons, maybe vouch them, and a way to update easily. But I guess that's not necessary to consider all of this immediately.

A small note about screen capture. I believe you could easily improve it by turning on android's preference option to display where you click/touch. This way, it'll become easier to follow what you are doing.

File: AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:52]: Let's say you decide that the preference key to recall what is activated is `addon_name+":activated"` and that by default an add-on is not activated. Then
`AnkiDroidApp.getSharedPrefs(Preferences.this).getBoolean(addon_name+":activated", false)` allow you to get the data
`SharedPreferences.Editor edit = AnkiDroidApp.getSharedPrefs(getBaseContext()).edit()` give you editor (a way to save new data)
`edit.putBoolean(my_addon_name:"activated", is_activated);` to state that the new value is `is_activated`
and
`edit.apply()` allow to save all new values simultaneously.

In src/main/java/com/ichi2/anki/Preferences.java you may want to look at "initSubscreen" and at the case ""com.ichi2.anki.prefs.custom_buttons": to get examlpe

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:52]: On second thought, you may want to use a Set<String> in the preference, containing the names of all activated add-ons.

Comment (infinyte7) [reply AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:52]: If name collide then all the addons with that name will be activated. So, may ID for addons will be good here.

**Comment by infinyte7** [comment]
This will greatly help in introducing some features of Anki addons that can be implemented using JavaScript for AnkiDroid using JS API. For example ```Card info during review```. But that need to added to all decks manully. 
I will implement the feature if file with ```.jsaddon``` extension valid for addons or not using their ```manifest.json``` file.

Also there will be some place where user can search and find AnkiDroid JS addons like https://addon-docs.ankiweb.net/

File: AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:52]: This lead to the questions: how do you decide which ID to give to an add-on?

Comment (infinyte7) [reply AnkiDroid/src/main/java/com/ichi2/anki/AddonsAdapter.java:52]: Now I have implemented as it supplied from addon developer with alphanumeric string.

**Comment by infinyte7** [comment]
demo

**Comment by infinyte7** [comment]
This is second commit to this PR. 
The addons containing ```{{Deck}}``` or ```{{Subdeck}}``` needs to converted to respective deckname. And similar transformation to be done if addons contains ```{{anything}}```

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 2]
```diff
@@ -2406,7 +2450,10 @@ public void fillFlashcard() {
             Timber.w("fillFlashCard() called with no card content");
             return;
         }
-        final String cardContent = mCardContent;
+
+        String addonsContent = getEnabledAddonsContent();
+        final String cardContent = mCardContent + addonsContent;
+
```
Comment (infinyte7) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:2456]: If ```addonsContent``` contains ```{{Deck}}``` then will functions will transform to respective deck name?
I have tried this but didn't work.
```
addonsContent = mCardTemplate.render(addonsContent, "", "");
```

Comment (infinyte7) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:2456]: I have tried following also but didn't work.
```
        String addonsContent = getEnabledAddonsContent();

        HashMap<String, String> context = new HashMap<>();
        Template template = new Template(addonsContent, context);
        addonsContent = template.render();
```

Getting ```{unknown field Deck}```

**Comment by infinyte7** [comment]
</img>

**Comment by infinyte7** [comment]
For importing addon, addon must contain two files.
```
content.html
manifest.json
```

content.html is file which get added to webview and manifest.json used for getting id.

The id in manifest.json and folder name of must be same.

The structure of ```js_addon_progress_bar_1234578```
```
js_addon_progress_bar_1234578
↳ content.css
↳ content.html
↳ content.js
↳ manifest.json
```
The file zipped with ```.jsaddon``` extension.

**Comment by mikehardy** [comment]
I'm sorry I haven't given this a real scan yet, but I wonder if it we could use the standard package.json format and thus rely on npmjs.com for distribution etc 🤔  ?

**Comment by infinyte7** [comment]
I haven't implemented any packages for npmjs so I will study the distribution. The format can be updated to package.json but for distribution I will read the documentation and then try to implement.

Distribution from npmjs will be better because there will no need for Anki Ecosystem to view and manage those files. That totally rely on developer. I will view the documentation of npmjs for better implementation. :)

**Comment by mikehardy** [comment]
the package.json format is definitely the most well known. It defines an entry point for the package, naming (and npmjs.com will enforce unique names :-)) plus I think you can tag on npmjs.com with keywords so we could do a search or something for "ankidroid-js-addon" or something if we make a convention for that. It can also define homepages and version and dependencies (regular ones, devDependencies, peerDependencies) etc.

For our purposes we likely won't be using anything but the most basic keywords in the package.json format but it keeps us using standard things which is nice

**Comment by infinyte7** [comment]
I understood it. I will change the manifest.json to package.json and will use useful and needed keywords from package.json.

**Comment by infinyte7** [comment]
**Implementation**
1. Published addon as npm package to npmjs
e.g. https://www.npmjs.com/package/ankidroid-js-addon-progress-bar

2. In AnkiDroid downloading latest package info as json file from
```http://registry.npmjs.org/{ file name from edit text }/latest ```
e.g. http://registry.npmjs.org/ankidroid-js-addon-progress-bar/latest 

     I haven't found way to download npm package directly.
     https://stackoverflow.com/questions/50933603/download-npm-dist-package-without-having-to-install-npm

3. Reading json file and getting ```tarball``` link to download the package

4. The npm package file downloded with ```.tgz```
e.g. ```ankidroid-js-addon-progress-bar.tgz```

5. Decompressing the tar file to addons folder using https://github.com/thrau/jarchivelib

6. In AbstractFlashcardViewer.java, the ```index.js``` read from ```addons/ankidroid-js.../package/index.js``` and added to reviewer

**Demo**

**Comment by infinyte7** [comment]
> For importing addon, addon must contain two files.
> 
> ```
> content.html
> manifest.json
> ```
> 
> content.html is file which get added to webview and manifest.json used for getting id.
> 
> The id in manifest.json and folder name of addon must be same.
> 
> The structure of `js_addon_progress_bar_1234578`
> 
> ```
> js_addon_progress_bar_1234578
> ↳ content.css
> ↳ content.html
> ↳ content.js
> ↳ manifest.json
> ```
> 
> In this case addon folder name and id both are `js_addon_progress_bar_1234578`.
> 
> The folder `js_addon_progress_bar_1234578` and files zipped with `.jsaddon` extension.
> 
> For testing I attached two jsaddons
> 
> [js_addons_progress_bar_12345678.zip](https://github.com/ankidroid/Anki-Android/files/5750872/js_addons_progress_bar_12345678.zip)
> 
> [js_addons_top_bar_12345678.zip](https://github.com/ankidroid/Anki-Android/files/5750873/js_addons_top_bar_12345678.zip)
> 
> Change the extension from zip to `jsaddon`.

This is outdated now.

**Comment by mikehardy** [comment]
I have still been booked up completely on some contract work but every time I see one of your animated gifs it is 🤤 ! this looks very powerful

**Comment by infinyte7** [comment]
The information need for contract is stored in package.json file. ```Email, ankidroid_api, author```
If ankidroid_api does not match api in AnkiDroid then ```.tgz``` file will not downloaded. So checking ```email, ankidroid_api, keywords``` in package.json will stop from downloading other npmjs package. 

I think, gifs is better than written description for explaining workflow of the PR. As it changes ui element also. :)

**Comment by infinyte7** [comment]
Also I think that one file in wiki should list all the addons that has been added to npmjs by a developer like [wiki/Third-Party-Apps](https://github.com/ankidroid/Anki-Android/wiki/Third-Party-Apps). New addons can be listed by the addon developer.

**Comment by mikehardy** [comment]
Is there way to do "discovery by convention" with a search powered by someone else?
What I mean concretely is: what if we had a convention of specific tags/keywords you should use when you upload to npmjs.com, then for JS addons we had a wiki page that had the npmjs.com search for those keywords in it, and we used the same search in the app?

That way all the search/discovery would be offloaded to npmjs.com and add on authors? zero maintenance burden unless npmjs.com changed their URL query interface...

**Comment by infinyte7** [comment]
I found this url useful.
https://www.npmjs.com/search?q=keywords:ankidroid-js-addon

In app we may add option for Get more addons.
1. It will prompt for opening link in browser. (In wiki a line should be mentioned for addon developer to add keyword for discoverability, like ```ankidroid-js-addon```)
https://www.npmjs.com/search?q=keywords:ankidroid-js-addon

2. Users view the listed addons on npmjs and select any addon
3. Copy their ```npm i ankidroid-js-addon-...``` and paste it in AnkiDroid->Addons

**Comment by infinyte7** [comment]
Implementation
1. It will download latest info of addon file
2. Check for following info if present in .json file

```
name
version 
author
ankidroid_js_api
homepage
keywords
```

If in latest json file, ```ankidroid_js_api``` is equal to 0.0.1 (current js api version) and ```keywords``` is equals to ```ankidroid-js-addon``` then get tarball link for downloading addon.tgz file and extract then copy it to AnkiDroid/addons folder.

https://github.com/ankidroid/Anki-Android/pull/7958/files#diff-e71d85656c47c1c4a8f0135d9b62eb6db857881841574c79522848f29725386dR348-R375

**Comment by infinyte7** [comment]
Implementation
1. Added addon type, ```reviewer``` and ```note_editor```
2. The conrent of addon with ```reviewer``` type will get added to reviewer from js file wrapped in ```<script>``` tag
3. The addon with ```note_editor``` type to be implemented.

Idea:
The javascript function can be called using https://github.com/evgenyneu/js-evaluator-for-android/ . So, in note editor a button will get added to toolbar with tag containing addon's name. When the toolbar button pressed, it will take ```selected text``` from FieldEditText and pass to jsEvaluate java function. The jsEvaluate java function will get tag of clicked toolbar button and open js file with name equal to tag of button and pass the selected text to addon's function implementation. The addon's js function will return the manipulated string to FieldEditText in note editor. Even if other implementation can be done without selected text as defined in js file in addon's folder. This will implemented in separated PR.

demo_4

**Comment by infinyte7** [comment]
**Update 1: Change addon info popup layout**
</img>

**Update 2: Remove prefs when addon delete**

**Update 3: Added a new addon for button and card counts to [npmjs](https://www.npmjs.com/package/ankidroid-js-addon-button-card-count)**
</img>

**Comment by infinyte7** [comment]
1. JS Addons in Advance settings
2. Layout for showing message to download addons when no addons added
3. Open npmjs.org in app

</img> </img> </img>

File: AnkiDroid/src/test/java/com/ichi2/anki/AddonsBrowserTest.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.ichi2.anki;
+
+public class AddonsBrowserTest extends RobolectricTest {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/test/java/com/ichi2/anki/AddonsBrowserTest.java:3]: What's the point of an empty class?

**Review by Arthur-Milchior — COMMENTED** [review]
Are 22 commits really useful ? Hard to review from scratch right now. For example, I'd like "first commit" to explain a little bit more what I'm reading. I don't know NPM at all, I don't really know JS, so I need commit message to guide me through what's going on to be able to review this

File: AnkiDroid/src/test/java/com/ichi2/anki/AddonsBrowserTest.java (hunk 3 above)
Comment (infinyte7) [reply AnkiDroid/src/test/java/com/ichi2/anki/AddonsBrowserTest.java:3]: To pass build of project.

**Comment by infinyte7** [comment]
I think it is not good idea to review from `First commit` because there are lot of changes done.
Should I squash all commit in to one?

**Comment by infinyte7** [comment]
I am creating new PR for this with detailed commit message.

**Comment by Arthur-Milchior** [comment]
Squashing everything will not really help me. I can not really read 1157 new lines of code and check that I'm okay with all of them. It's particularly hard because I don't know anything about JS and NPM and other tool you use.

It is possible that you actually can not split this into smaller commits that all makes sens by themselves; it sometime happens and makes things hard. However, ideally, what I would love to see is a story. You make a first commit that compiles, pass tests, with a commit message that explains to me and future reader why you do it, why it is useful. Then you add another part of the code etc... 
Using `git rebase -i` (for interactive) you usually can change history, modify a commit to correct what was added in it.
Of course, it works best if you do it progressively through the history of the PR; rewritting all of this as a consistent story later requires more thought process.

**Comment by infinyte7** [comment]
I have created separate PR because creating is easy and better than rewriting history. :)
Also I have tried to explain the working of js addons idea.

**Pull request closed without merging**